doc/development/autograph.rst

```diff
@@ -0,0 +1,743 @@
+...     if x > 5:
+...         y = "a"
+...     else:
+...         y = "b"
+...     return y
+>>> plxpr = make_plxpr(f)(0.5)
+>>> eval_jaxpr(plxpr.jaxpr, plxpr.consts, 7.0)
+TypeError: Value 'a' with type <class 'str'> is not a valid JAX type
+
+For loops
+---------
+
+Most ``for`` loop constructs will be properly captured and compiled by AutoGraph.
+
+.. code-block:: python
+
+    dev = qml.device("default.qubit", wires=1)
+
+    @qml.qnode(dev)
+    def f():
+        for x in [0, 1, 2]:
+            qml.RY(x * jnp.pi / 4, wires=0)
+        return qml.expval(qml.PauliZ(0))
+
+>>> plxpr = make_plxpr(f)()
+>>> eval_jaxpr(plxpr.jaxpr, jaxpr.consts)
+[Array(-0.70710678, dtype=float64)]
+
+This includes automatic unpacking and enumeration through JAX arrays:
+
+>>> def f(weights):
+...     z = 0.
+...     for i, (x, y) in enumerate(weights):
+...         z = i * x + i ** 2 * y
+...     return z
+>>> weights = jnp.array([[0.1, 0.2, 0.3, 0.4], [0.5, 0.6, 0.7, 0.8]]).T
+>>> plxpr = make_plxpr(f)(weights)
+>>> eval_jaxpr(plxpr.jaxpr, [], weights)
+Array(8.4, dtype=float64)
+
+The Python ``range`` function is also supported by AutoGraph, even when
+its input is a **dynamic variable** (i.e., its numeric value is only known at
+runtime):
+
+>>> def f(n):
+...     x = -jnp.log(n)
+...     for k in range(1, n + 1):
+...         x = x + 1 / k
+...     return x
+>>> plxpr = make_plxpr(f)(0)
+>>> eval_jaxpr(plxpr.jaxpr, plxpr.consts, 1000)
+[Array(0.57771558, dtype=float64, weak_type=True)]
+
+Indexing within a loop
+~~~~~~~~~~~~~~~~~~~~~~
+
+Indexing arrays within a for loop will generally work, but care must be taken.
+
+For example, using a for loop with static bounds to index a JAX array is straightforward:
+
+>>> dev = qml.device("default.qubit", wires=3)
+... @qml.qnode(dev)
+... def f(x):
+...     for i in range(3):
+...         qml.RX(x[i], wires=i)
+...     return qml.expval(qml.PauliZ(0))
+>>> weights = jnp.array([0.1, 0.2, 0.3])
+>>> plxpr = make_plxpr(f)(weights)
+>>> eval_jaxpr(plxpr.jaxpr, plxpr.consts, weights)
+[Array(0.99500417, dtype=float64)]
+
+However, indexing within a for loop with AutoGraph will require that the object indexed is
+a JAX array or dynamic runtime variable.
+
+If the array you are indexing within the for loop is not a JAX array
+or dynamic variable, an error will be raised:
+
+>>> @qml.qnode(dev)
+... def f():
+...     x = [0.1, 0.2, 0.3]
+...     for i in range(3):
+...         qml.RX(x[i], wires=i)
+...     return qml.expval(qml.PauliZ(0))
+>>> plxpr = make_plxpr(f)()
+AutoGraphError: Tracing of an AutoGraph converted for loop failed with an exception:
+  TracerIntegerConversionError:    The __index__() method was called on traced array with shape int64[]
+    The error occurred while tracing the function functional_for at /Users/lillian.frederiksen/pennylane/pennylane/capture/autograph/ag_primitives.py:176 for jit. This concrete value was not available in Python because it depends on the value of the argument i.
```

Comment:
Just a heads up that your name is showing up in the error message here as well. 😄

Reply:
Hahah, thanks, I'll truncate this a bit 😅